rust/src/dcerpc/dcerpc.rs

```diff
@@ -924,36 +820,83 @@ impl DCERPCState {
         }
         // Check if header data was complete. In case of EoF or incomplete data, wait for more
         // data else return error
-        if self.header.is_none() && !cur_i.is_empty() {
-            parsed = self.process_header(cur_i);
-            if parsed == -1 {
-                return AppLayerResult::incomplete(0, DCERPC_HDR_LEN as u32);
-            }
-            if parsed < 0 {
-                return AppLayerResult::err();
+        let dir_header = if direction == Direction::ToServer { &self.header_ts } else { &self.header_tc };
+        if dir_header.is_none() {
+            match parser::parse_dcerpc_header(cur_i) {
+                Ok((_leftover_bytes, header)) => {
+                    if header.rpc_vers != 5
+                        || (header.rpc_vers_minor != 0 && header.rpc_vers_minor != 1)
+                    {
+                        SCLogDebug!(
+                            "DCERPC Header did not validate. Major version: {:?} Minor version: {:?}",
+                            header.rpc_vers,
+                            header.rpc_vers_minor
+                        );
+                        return AppLayerResult::err();
+                    }
+                    if direction == Direction::ToServer {
+                        self.header_ts = Some(header);
+                    } else {
+                        self.header_tc = Some(header);
+                    }
+                }
+                Err(Err::Incomplete(_)) => {
+                    // Insufficient data.
+                    SCLogDebug!("Insufficient data while parsing DCERPC header");
+                    return AppLayerResult::incomplete(0, DCERPC_HDR_LEN as u32);
+                }
+                Err(Err::Error(Error{code:ErrorKind::Eof, ..})) => {
+                    SCLogDebug!("EoF reached while parsing DCERPC header");
+                    return AppLayerResult::incomplete(0, DCERPC_HDR_LEN as u32);
+                }
+                Err(_) => {
+                    // Error, probably malformed data.
+                    SCLogDebug!("An error occurred while parsing DCERPC header");
+                    return AppLayerResult::err();
+                }
             }
         } else {
             frag_bytes_consumed = DCERPC_HDR_LEN;
         }
+        let parsed = DCERPC_HDR_LEN;
+        let fraglen = if direction == Direction::ToServer {
+            if let Some(h) = &self.header_ts {
+                h.frag_length
+            } else {
+                0
+            }
+        } else {
+            if let Some(h) = &self.header_tc {
+                h.frag_length
+            } else {
+                0
+            }
+        };
 
-        let fraglen = self.get_hdr_fraglen().unwrap_or(0);
-
-        if cur_i.len() < (fraglen - frag_bytes_consumed) as usize {
+        if (cur_i.len() + frag_bytes_consumed as usize) < fraglen as usize {
```

Comment:
This does appear to restore correct behavior.

Reply:
@inashivb what do you think of the behavior when fraglen < 16 ?

I think we should set an event

Reply:
> @inashivb what do you think of the behavior when fraglen < 16 ?
> 
> I think we should set an event

Agreed.

Reply:
But DCERPC has no event yet...